tardis/energy_input/util.py

```diff
@@ -0,0 +1,333 @@
+    x : float64
+        calculated vector x position
+    y : float64
+        calculated vector y position
+    z : float64
+        calculated vector z position
+    """
+
+    def __init__(self, r, theta, phi=0.0):
+        self.r = r
+        self.theta = theta
+        self.phi = phi
+
+    @property
+    def mu(self):
+        return np.cos(self.theta)
+
+    @property
+    def cartesian_coords(self):
+        # 0.5*np.pi subtracted because of the definition of theta
+        # in astropy.coordinates.cartesian_to_spherical
+        x, y, z = spherical_to_cartesian(
+            self.r, self.theta - 0.5 * np.pi, self.phi
+        )
+        return x.value, y.value, z.value
+
+
+def kappa_calculation(energy):
+    """
+    Calculates kappa for various other calculations
+    i.e. energy normalized to electron rest energy
+    511.0 KeV
+
+    Parameters
+    ----------
+    energy : float
+
+    Returns
+    -------
+    kappa : float
+
+    """
+    k = energy / 511.0
+    return k
+
+
+def euler_rodrigues(theta, direction):
+    """
+    Calculates the Euler-Rodrigues rotation matrix
+
+    Parameters
+    ----------
+    theta : float
+    direction : SphericalVector object
+
+    Returns
+    -------
+    rotation matrix : Two dimensional Numpy array, dtype float
+
+    """
+    a = np.cos(theta / 2)
+    dir_x, dir_y, dir_z = direction
+    b = dir_x * np.sin(theta / 2)
+    c = dir_y * np.sin(theta / 2)
+    d = dir_z * np.sin(theta / 2)
+
+    er11 = a ** 2.0 + b ** 2.0 - c ** 2.0 - d ** 2.0
+    er12 = 2.0 * (b * c - a * d)
+    er13 = 2.0 * (b * d + a * c)
+
+    er21 = 2.0 * (b * c + a * d)
+    er22 = a ** 2.0 + c ** 2.0 - b ** 2.0 - d ** 2.0
+    er23 = 2.0 * (c * d - a * b)
+
+    er31 = 2.0 * (b * d - a * c)
+    er32 = 2.0 * (c * d + a * b)
+    er33 = a ** 2.0 + d ** 2.0 - b ** 2.0 - c ** 2.0
+
+    return np.array(
+        [[er11, er12, er13], [er21, er22, er23], [er31, er32, er33]]
+    )
+
+
+def solve_quadratic_equation(x, y, z, x_dir, y_dir, z_dir, radius_velocity):
+    """
+    Solves the quadratic equation for the distance to the shell boundary
+
+    Parameters
+    ----------
+    x,y,z : float
+    x_dir, y_dir, z_dir : float
+    radius_velocity : float
+
+    Returns
+    -------
+    solution_1 : float
+    solution_2 : float
+
+    """
+    b = 2.0 * (x * x_dir + y * y_dir + z * z_dir)
+    c = -(radius_velocity ** 2) + x ** 2 + y ** 2 + z ** 2
+    root = b ** 2 - 4 * c
+    solution_1 = -np.inf
+    solution_2 = -np.inf
+    if root > 0.0:
+        solution_1 = 0.5 * (-b + np.sqrt(root))
+        solution_2 = 0.5 * (-b - np.sqrt(root))
+    elif root == 0:
+        solution_1 = -0.5 * b
+    return solution_1, solution_2
+
+
+def klein_nishina(energy, theta_C):
+    """
+    Calculates the Klein-Nishina equation
+
+    Parameters
+    ----------
+    energy : float
+        Packet energy
+    theta_C : float
+        Compton angle
+    Returns
+    -------
+    float
+        Klein-Nishina solution
```

Comment:
maybe a link to wikipedia or something where we can see the formula - definitely add in latex to the doc string